pygsti/tools/lindbladtools.py

```diff
@@ -72,35 +84,158 @@ def create_elementary_errorgen_dual(typ, p, q=None, sparse=False, normalization_
-        elem_errgen[:, i] = rho1.ravel()
-        # ^ That line used to branch depending on the value of "sparse", but it
-        #   turns out that both codepaths produced the same result.
+    #if p or q is a sparse matrix fall back to original implementation
+    if not isinstance(p, _np.ndarray) or (q is not None and not isinstance(q, _np.ndarray)):
+        for i, rho0 in enumerate(basis_matrices('std', d2)):  # rho0 == input density mx
+            # Only difference between H/S/C/A is how they transform input density matrices
+            if typ == 'H':
+                rho1 = -1j * (p @ rho0 - rho0 @ p)  # -1j / (2 * d2) *
+            elif typ == 'S':
+                rho1 = (p @ rho0 @ pdag)  # 1 / d2 *
+            elif typ == 'C':
+                rho1 = (p @ rho0 @ qdag + q @ rho0 @ pdag)  # 1 / (2 * d2) *
+            elif typ == 'A':
+                rho1 = 1j * (p @ rho0 @ qdag - q @ rho0 @ pdag)  # 1j / (2 * d2)
+            elem_errgen[:, i] = rho1.flatten()[:, None] if sparse else rho1.flatten()
+    else:
+        # Loop through the standard basis as all possible input density matrices
+        for i in range(d):  # rho0 == input density mx
+            for j in range(d):
+                # Only difference between H/S/C/A is how they transform input density matrices
+                if typ == 'H':
+                    rho1 = _np.zeros((d,d), dtype=_np.complex128)
+                    rho1[:, j] = -1j*p[:, i]
+                    rho1[i, :] += 1j*p[j, :]
+                elif typ == 'S':
+                    rho1 = p[:,i].reshape((d,1))@pdag[j,:].reshape((1,d))
+                elif typ == 'C':
+                    rho1 = p[:,i].reshape((d,1))@qdag[j,:].reshape((1,d)) + q[:,i].reshape((d,1))@pdag[j,:].reshape((1,d))
+                elif typ == 'A':
+                    rho1 = 1j*(p[:,i].reshape((d,1))@ qdag[j,:].reshape((1,d))) - 1j*(q[:,i].reshape((d,1))@pdag[j,:].reshape((1,d)))
+
+                elem_errgen[:, d*i+j] = rho1.flatten()[:, None] if sparse else rho1.flatten()
 
     return_normalization = bool(normalization_factor == 'auto_return')
     if normalization_factor in ('auto', 'auto_return'):
         primal = create_elementary_errorgen(typ, p, q, sparse)
         if sparse:
-            normalization_factor = _np.vdot(elem_errgen.toarray(), primal.toarray())
+            normalization_factor = _np.vdot(elem_errgen.toarray().flatten(), primal.toarray().flatten())
         else:
-            normalization_factor = _np.vdot(elem_errgen, primal)
+            normalization_factor = _np.vdot(elem_errgen.flatten(), primal.flatten())
     elem_errgen *= _np.real_if_close(1 / normalization_factor).item()  # item() -> scalar
+    if sparse: elem_errgen = elem_errgen.tocsr()
+    return (elem_errgen, normalization_factor) if return_normalization else elem_errgen
+
+#TODO: Should be able to leverage the structure of the paulis as generalized permutation
+#matrices to avoid explicitly doing outer products
```

Comment:
You noted this in your PR too, but want to make sure we don't lose track of it - another thing to add to 2.0 or spawn off as a separate issue?